compiler/src/dotty/tools/dotc/core/TypeOps.scala

```diff
@@ -429,17 +430,22 @@ object TypeOps:
             case _ =>
               emptyRange // should happen only in error cases
           }
-        case tp: ThisType if toAvoid(tp.cls) =>
-          range(defn.NothingType, apply(classBound(tp.cls.classInfo)))
+        case tp: ThisType =>
+          tp
         case tp: SkolemType if partsToAvoid(Nil, tp.info).nonEmpty =>
           range(defn.NothingType, apply(tp.info))
         case tp: TypeVar if mapCtx.typerState.constraint.contains(tp) =>
           val lo = TypeComparer.instanceType(
             tp.origin, fromBelow = variance > 0 || variance == 0 && tp.hasLowerBound)(using mapCtx)
           val lo1 = apply(lo)
           if (lo1 ne lo) lo1 else tp
-        case tp: LazyRef if isExpandingBounds =>
-          emptyRange
+        case tp: LazyRef =>
+          if locals.contains(tp.ref) then tp
```

Comment:
Why is it safe to return tp here, doesn't it need to be avoided if it contains a local symbol?

Reply:
Here `locals` mean locally bound `ParamRef` in a HKTypeLambda -- we remember that in the set `locals`. The set is not the set of local symbols to be avoided.

Reply:
Oh I see, would be good to give it another name to avoid confusing it with `localSyms` I think :). Maybe `localParamRefs` ?